Mathlib/ModelTheory/Basic.lean

```diff
@@ -0,0 +1,1100 @@
+
+/-- The empty language has no symbols. -/
+protected def empty : Language :=
+  ⟨fun _ => Empty, fun _ => Empty⟩
+#align first_order.language.empty FirstOrder.Language.empty
+
+instance : Inhabited Language :=
+  ⟨Language.empty⟩
+
+/-- The sum of two languages consists of the disjoint union of their symbols. -/
+protected def sum (L : Language.{u, v}) (L' : Language.{u', v'}) : Language :=
+  ⟨fun n => Sum (L.Functions n) (L'.Functions n), fun n => Sum (L.Relations n) (L'.Relations n)⟩
+#align first_order.language.sum FirstOrder.Language.sum
+
+variable (L : Language.{u, v})
+
+/-- The type of constants in a given language. -/
+-- Porting note: The linter does not exist yet, according to Algebra.Hom.GroupAction.
+-- @[nolint has_nonempty_instance]
+protected def Constants :=
+  L.Functions 0
+#align first_order.language.constants FirstOrder.Language.Constants
+
+@[simp]
+theorem constants_mk₂ (c f₁ f₂ : Type u) (r₁ r₂ : Type v) :
+    (Language.mk₂ c f₁ f₂ r₁ r₂).Constants = c :=
+  rfl
+#align first_order.language.constants_mk₂ FirstOrder.Language.constants_mk₂
+
+/-- The type of symbols in a given language. -/
+-- Porting note: The linter does not exist yet, according to Algebra.Hom.GroupAction.
+-- @[nolint has_nonempty_instance]
+def Symbols :=
+  Sum (Σl, L.Functions l) (Σl, L.Relations l)
+#align first_order.language.symbols FirstOrder.Language.Symbols
+
+/-- The cardinality of a language is the cardinality of its type of symbols. -/
+def card : Cardinal :=
+  #L.Symbols
+#align first_order.language.card FirstOrder.Language.card
+
+/-- A language is relational when it has no function symbols. -/
+class IsRelational : Prop where
+  /-- There are no function symbols in the language. -/
+  empty_functions : ∀ n, IsEmpty (L.Functions n)
+#align first_order.language.is_relational FirstOrder.Language.IsRelational
+
+/-- A language is algebraic when it has no relation symbols. -/
+class IsAlgebraic : Prop where
+  /-- There are no relation symbols in the language. -/
+  empty_relations : ∀ n, IsEmpty (L.Relations n)
+#align first_order.language.is_algebraic FirstOrder.Language.IsAlgebraic
+
+variable {L} {L' : Language.{u', v'}}
+
+theorem card_eq_card_functions_add_card_relations :
+    L.card =
+      (Cardinal.sum fun l => Cardinal.lift.{v} (#L.Functions l)) +
+        Cardinal.sum fun l => Cardinal.lift.{u} (#L.Relations l) := by
+  simp [card, Symbols]
+#align
+  first_order.language.card_eq_card_functions_add_card_relations
+  FirstOrder.Language.card_eq_card_functions_add_card_relations
+
+instance [L.IsRelational] {n : ℕ} : IsEmpty (L.Functions n) :=
+  IsRelational.empty_functions n
+
+instance [L.IsAlgebraic] {n : ℕ} : IsEmpty (L.Relations n) :=
+  IsAlgebraic.empty_relations n
+
+instance isRelational_of_empty_functions {symb : ℕ → Type _} :
+    IsRelational ⟨fun _ => Empty, symb⟩ :=
+  -- Porting note: Previous code was
+  -- ⟨fun _ => Empty.isEmpty⟩
+  --
+  -- No instance Empty.isEmpty.
```

Comment:
It will be easier on reviewers if you quote the errors exactly. This was `unknown constant`, not `No instance` (which would have been weird).

Reply:
In mathlib3 we were using an automatically named instance, which changed names. You can diagnose these sorts of issues by opening the mathlib3 file, and using jump to definition on the original identifier.

Reply:
I've fixed this, and the following three similar porting notes.

Reply:
Sorry, it seems I didn't actually commit this...